maven-plugin-tools-annotations/src/main/java/org/apache/maven/tools/plugin/extractor/annotations/scanner/MojoAnnotationsScanner.java

```diff
@@ -37,13 +37,21 @@
 {
     String ROLE = MojoAnnotationsScanner.class.getName();
 
+    String V4_API_PLUGIN_PACKAGE = "org.apache.maven.api.plugin";
```

Comment:
Maybe a bit late but why are the annotations part of maven-api-core? IMHO those are used exclusively at build time (not at run time) by m-plugin-tools only to build the descriptor, so why defining them globally?
Also why can't the old annotations be extended with just that a boolean flag for v4Api?

Reply:
> Maybe a bit late but why are the annotations part of maven-api-core? IMHO those are used exclusively at build time (not at run time) by m-plugin-tools only to build the descriptor, so why defining them globally?

The plugin annotations are the first classes needed for people that consume the maven api.  It's part of the API and I would not consider the fact that they are only used at build time (that's an implementation detail of how plugin works).  That is also related to the fact that the `MojoDescriptor` is currently not exposed in the api.  So there's a small change in semantics, as I consider the annotations as _the api_ for plugin authors, while the descriptor is kept an implementation detail (which is good imho as it inherit plexus classes which should be hidden).
Some attempts have been made over the years to define plugins in other languages such as the scripting stuff, but it's being deprecated and will be removed soon I think.  I think that's an indication that (nearly) all plugins are simply written using the provided annotations, so they _de facto_ constitute the mojo API.

> Also why can't the old annotations be extended with just that a boolean flag for v4Api?

They could.  Bur for the reason above, I think they should belong to the "_maven api_" which I think is best kept in a single location rather than splitted across various modules and projects.  Now, people can simply depend on the `maven-api-core` jar and have a clear view of all the maven api they can use.

Reply:
> the annotations as the api for plugin authors

Agree, when they use the m-plugin-plugin for building their plugin (which is likely but not a must).
Still mixing runtime API and build time API in the same artifact is confusing to me.

> while the descriptor is kept an implementation detail (which is good imho as it inherit plexus classes which should be hidden).

Hmm, let us call it SPI at least as it defines the contract between Maven and its Plugins.

While I agree with most of your statements I still don't understand why those annotations need to be part of Maven rather than part of the (anyhow used) m-plugin tools. They are exclusively used there!
Also they inflate/pollute the Maven classpath for no reason as they are `RetentionPolicy.CLASS`. Mixing those different levels of API in a single artifact is not a very good design IMHO.

Reply:
I disagree that the fact they are not retained at runtime implies they should be in a different jar.  Should we change the retention in the new api ? ;-) 

More seriously, I agree they're not used at runtime, but I think from a user facing pov, they are the API for mojos, at the same level than the `Mojo` interface. The fact they are interpreted at compile time and translated into an xml is mainly an implementation detail for the average user. 
I think having a monolithic jar for the API  makes it way easier to just depend on the `maven-core-api` without having to look for additional jars, depending if you want to do simple mojos, some access to the local repository, some custom dependency collection or whatever part which has been moved to external components.  We don't even know which ones are deprecated and which ones should be used (for dependency collection / resolution).  With a single jar, everything coming from this jar (and it's dependencies) is part of the new API, everything else isn't.